verilog/preprocessor/verilog_preprocess.h

```diff
@@ -120,6 +145,60 @@ class VerilogPreprocess {
 
   void RegisterMacroDefinition(const MacroDefinition&);
 
+  const Config config_;
+
+  // State of a block that can have a sequence of sub-blocks with conditions
+  // (ifdef/elsif/else/endif). Only the first of the subblock whose condition
+  // matches will be selected - or the else block if none before matched.
+  class BranchBlock {
+   public:
+    BranchBlock(bool is_enabled, bool condition,
+                const verible::TokenInfo& token)
+        : outer_scope_enabled_(is_enabled), branch_token_(token) {
+      UpdateCondition(token, condition);
+    }
+
+    // Is the current block selected, i.e. should tokens pass through.
+    bool InSelectedBranch() const {
+      return outer_scope_enabled_ && current_branch_condition_met_;
+    }
+
+    // Update condition e.g. in `elsif. Return 'false' if already in an
+    // else clause.
+    bool UpdateCondition(const verible::TokenInfo& token, bool condition) {
+      if (in_else_) return false;
+      branch_token_ = token;
+      // Only the first in a row of matching conditions will select block.
+      current_branch_condition_met_ = !any_branch_matched_ && condition;
+      any_branch_matched_ |= condition;
+      return true;
+    }
+
+    // Start an `else block. Uses its internal state to determine if
+    // this will put is InSelectedBranch().
+    // Returns 'false' if already in an else block.
+    bool StartElse(const verible::TokenInfo& token) {
+      if (in_else_) return false;
+      in_else_ = true;
+      branch_token_ = token;
+      current_branch_condition_met_ = !any_branch_matched_;
+      return true;
+    }
+
+    const verible::TokenInfo& token() const { return branch_token_; }
+
+   private:
+    const bool outer_scope_enabled_;
+    verible::TokenInfo branch_token_;  // FYI for error reporting.
+    bool any_branch_matched_ = false;  // only if no branch, `else will
+    bool in_else_ = false;
```

Comment:
Is `branch_token_` always an ifdef or ifndef?
What if each level of the stack tracked up to two locations:
* branch open point (ifdef/ifndef)
* current clause start (could be any of ifdef/ifndef/elsif/else)

That could effectively tell you `any_branch_matched_` and `in_else_`, and would also provide good reference points for diagnostics, e.g. _"error in else clause at <location> of branch started at <location>"_.

Reply:
right now, 'branch_token_' is always the last token so can be one of ifdef/ifndef/else/elsif. So the diagnostic right now is already referring to the last relevant point.

But good idea for the future in case we need to report ambiguous situations. Ideally, I'd like to expand  the error message to have 'reference' messages (right now I report errors with reference as two independent error messages)